chainstate/src/detail/chainstateref.rs

```diff
@@ -292,17 +303,22 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
         let id = self
             .get_block_id_by_height(height)?
             .ok_or(PropertyQueryError::BlockForHeightNotFound(*height))?;
+        let id = id.classify(self.chain_config).block_id().expect("Genesis has no header");
```

Comment:
The reason why I'm excited for separating genesis block from chain blocks is to avoid these expects as much as possible... we have to be really careful. If p2p, for example, tries to get the header from a height 0 for any reason because a peer sends some fishy information that they can manipulate, the result should not be crashing the node. Generally speaking, in the cases of property querying, it's always preferred to return `Ok(None)` in such cases unless it's provable that no one will ever call this function with an invalid height.

Reply:
Good points. The point of this function is to get block header from height. Returning `Ok(None)`, we would lose the distinction between asking about block header at height 0 and asking about block header at height beyond the current tip. Do we care about that? Probably not at the moment, since the expect is not hit in the tests but in general I'm not sure.

On a bit more high-level note, I think one of the reasons for bunch of remaining `expect`s is that not all types currently reflect the distinction between the genesis and a chain block. The prime offender currently being `BlockHeight`. Genesis corresponds to block height 0, chain block corresponds to a non-zero block height, and in general block height may refer to either genesis or a chain block.

So similar thing could be done to it as to other types in this PR, i.e. we could have a `BlockHeight` and a `NonZeroBlockHeight`, with some conversion functions:
```rust
BlockHeight::non_zero(Self) -> Option<NonZeroBlockHeight>
NonZeroBlockHeight::into(Self) -> BlockHeight
```

Then, there could be methods like:
```rust
get_block_id_by_height(NonZeroBlockHeight) -> Result<Id<Block>>
get_gen_block_id_by_height(BlockHeight) -> Result<Id<GenBlock>>
get_block_by_height(NonZeroBlockHeight) -> Result<Block>
// This function could then have signature, that makes it callable only with non-zero height:
get_header_from_height(NonZeroBlockHeight) -> Result<BlockHeader>
```

With this we would not have to choose between an expect and an error condition that the caller has to deal with.

Reply:
> Returning Ok(None), we would lose the distinction between asking about block header at height 0 and asking about block header at height beyond the current tip

Easy. Just return an error if the value is zero. That's alright too. 

> So similar thing could be done to it as to other types in this PR, i.e. we could have a BlockHeight and a NonZeroBlockHeight, with some conversion functions:

Then we probably have to expect on some things... I don't see this as a bad idea to be tried out, but maybe not in this PR. Maybe after this one you can give it a shot. I think this PR is already huge and experimenting further could lead to this becoming a long standing PR, which we want to avoid because this is currently blocking everything else.

Reply:
Btw: https://doc.rust-lang.org/std/num/struct.NonZeroU64.html

Reply:
> Then we probably have to expect on some things... I don't see this as a bad idea to be tried out, but maybe not in this PR. Maybe after this one you can give it a shot.

But hopefully there would be fewer of them... and possibly in less tricky places

> I think this PR is already huge and experimenting further could lead to this becoming a long standing PR, which we want to avoid because this is currently blocking everything else.

Agreed

Reply:
Added a new error for this for the time of being